digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/assetUpdater/ChangeReporter.scala

```diff
@@ -0,0 +1,231 @@
+      }
+
+    } else {
+      logger.warn("Asset does not have geometry")
+      ""
+    }
+  }
+
+  def getUrl: String = {
+    if (linearReference.nonEmpty) {
+      s"""$directLink#linkProperty/${linearReference.get.linkId}"""
+    }  else ""
+  }
+
+}
+
+sealed trait ChangeType {
+  def value: Int
+}
+
+object ChangeTypeReport {
+  
+  case object Creation extends ChangeType {
+    def value: Int = 1
+  }
+
+  case object Deletion extends ChangeType {
+    def value: Int = 2
+  }
+
+  case object Divided extends ChangeType {
+    def value: Int = 3
+  }
+
+  case object Replaced extends ChangeType {
+    def value: Int = 4
+  }
+  case object PropertyChange extends ChangeType {
+    def value: Int = 5
+  }
+  
+  /**
+    * For point asset
+    * */
+  case object Move extends ChangeType {
+    def value: Int = 7
+  }
+
+  /**
+    * For point asset
+    * */
+  case object Floating extends ChangeType {
+    def value: Int = 8
+  }
+}
+
+sealed trait ReportedChange {
+  def linkId: String
+  def changeType: ChangeType
+}
+
+case class AdministrativeClassChange(linkId: String, changeType: ChangeType, oldValue: Int, newValue: Option[Int]) extends ReportedChange
+case class TrafficDirectionChange(linkId: String, changeType: ChangeType, oldValue: Int, newValue: Option[Int]) extends ReportedChange
+case class RoadLinkAttributeChange(linkId: String, changeType: ChangeType, oldValues: Map[String, String], newValues: Map[String, String]) extends ReportedChange
+case class FunctionalClassChange(linkId: String, changeType: ChangeType, oldValue: Option[Int], newValue: Option[Int], source: String = "") extends ReportedChange
+case class LinkTypeChange(linkId: String, changeType: ChangeType, oldValue: Option[Int], newValue: Option[Int], source: String = "") extends ReportedChange
+
+
+/**
+  * 
+  * @param linkId     link where changes is happening TODO remove if not needed
+  * @param assetId    asset which is under samuutus, When there is more than one asset under samuutus (e.g merger or join) create new  [[ChangedAsset]] item for each asset.
+  * @param changeType characteristic of change
+  * @param before     situation before samuutus
+  * @param after      after samuutus
+  * */
+case class ChangedAsset(linkId: String, assetId: Long, changeType: ChangeType, before: Asset, after: Seq[Asset]) extends ReportedChange
+
+/**
+  *
+  * @param assetType
+  * @param changes
+  */
+case class ChangeReport(assetType: Int, changes: Seq[ReportedChange])
+
+object ChangeReporter {
+
+  lazy val awsService = new AwsService
+  lazy val s3Service: awsService.S3.type = awsService.S3
+  lazy val s3Bucket: String = Digiroad2Properties.samuutusReportsBucketName
+  implicit lazy val serializationFormats: Formats = DefaultFormats
+
+  private def getCSVRowForRoadLinkPropertyChanges(linkId: String, changeType: Int, changes: Seq[ReportedChange]) = {
+    val trafficDirectionChange = changes.find(_.isInstanceOf[TrafficDirectionChange])
```

Comment:
changes.groupBy(_.linkId).find(_.id==linkId) ja sitten for loop voi ehkä olla performanssi mielessä ehkä parempi. kun nyt joudutaan tekemään viisi loop. 

toi find tais hakea vaan ensimmäisen joka vastaa ehtoa, joten teoriassa toi vois ehkä valita väärän tiedon.

Reply:
ok